vcsm: Unify cache manipulating functions

Terminus-IMRC opened the pull request:
`vcsm` has four ioctl calls for cache flushing/cleaning/invalidating: `FLUSH`, `INVALID`, `CLEAN_INVALID`, `CLEAN_INVALID2`.
The operations are finally done by `dmac_{flush,clean,inv}_range()`, but the ioctl handlers used individual wrapper functions. This PR unifies them.

Terminus-IMRC committed:
vcsm: Define cache operation constants in user header

Without this change, users have to use raw values (1, 2, 3) to specify
cache operation.

Signed-off-by: Sugizaki Yukimasa <i.can.speak.c.and.basic@gmail.com>

Terminus-IMRC committed:
vcsm: Support for finding user/vc handle in memory pool

vmcs_sm_{usr,vc}_handle_from_pid_and_address() were failing to find
handle if specified user pointer is not exactly the one that the memory
locking call returned even if the pointer is in range of map/resource.
So fixed the functions to match the range.

Signed-off-by: Sugizaki Yukimasa <i.can.speak.c.and.basic@gmail.com>

Terminus-IMRC committed:
vcsm: Unify cache manipulating functions

Signed-off-by: Sugizaki Yukimasa <i.can.speak.c.and.basic@gmail.com>

Terminus-IMRC commented:
Oh, should I propose this PR to `rpi-4.14.y` branch?

6by9 commented on drivers/char/broadcom/vc_sm/vmcs_sm.c:
```diff
@@ -276,7 +276,10 @@ static unsigned int vmcs_sm_vc_handle_from_pid_and_address(unsigned int pid,
-			if (map->res_pid != pid || map->res_addr != addr)
+			if (map->res_pid != pid)
+				continue;
+			if (!(map->res_addr <= addr &&
```
Seems a little obfuscated.
To my mind 
```
if (addr < map->res_addr || addr > (map->res_addr + map->resource->res_size))
   continue;
```
is more obvious.
(In a percentage of cases it is likely to be faster too, as it can avoid computing ```addr > (map->res_addr + map->resource->res_size)``` if ```addr < map->res_addr``` was true)

Terminus-IMRC replied:
Yeah, mine was obscure and not efficient. Fixed in https://github.com/raspberrypi/linux/pull/2333/commits/872f3ca2c4dd268bc6d537151435e45f16224b0e.

Terminus-IMRC replied:
(I changed the second condition of yours from `addr > (map->res_addr + map->resource->res_size)` to `addr >= (map->res_addr + map->resource->res_size)` because yours may match an neighboring map area.)

6by9 commented on include/linux/broadcom/vmcs_sm_ioctl.h:
```diff
@@ -79,6 +79,11 @@ enum vmcs_sm_cache_e {
 	VMCS_SM_CACHE_BOTH,
 };
 
+/* Cache functions */
```
Could you add either a comment that this relates to struct vmcs_sm_ioctl_clean_invalid cmd and struct vmcs_sm_ioctl_clean_invalid2 invalidate_mode, or move it to just before those structs?

Terminus-IMRC replied:
Thanks. I did both: https://github.com/raspberrypi/linux/pull/2333/commits/6e3cd4cb50ec8f19287a5b00ddd4563c6a6e255c

6by9 commented on drivers/char/broadcom/vc_sm/vmcs_sm.c:
```diff
@@ -1256,61 +1256,106 @@ static const struct vm_operations_struct vcsm_vm_ops = {
 
-	if (addr >= end)
-		return;
+	if (block_size <= 0) {
```
Nit pick:
```if (!block_size)```
block_size is size_t = unsigned int, therefore it can never be < 0. Your comment already reflects this.

Terminus-IMRC replied:
I was writing the condition so that it can detect negative size if the size was unintentionally signed, but it's obviously unsigned here, so fixed in https://github.com/raspberrypi/linux/pull/2333/commits/872f3ca2c4dd268bc6d537151435e45f16224b0e.

6by9 commented on drivers/char/broadcom/vc_sm/vmcs_sm.c:
```diff
@@ -3006,36 +2985,12 @@ static long vc_sm_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
+							op->block_count, op->block_size,
+							op->inter_block_stride, op->invalidate_mode);
+					if (ret)
+						goto out;
```
You leak ```block``` if you goto out here. break is fine.

(I have a cleanup pending to this file that removes a load of the ```goto out``` calls, as they're only doing the equivalent of running on, or break within the switch statement. I'll let your PR go through first, as long as it is functionally correct).

Terminus-IMRC replied:
Thanks. It really leaks memory. Fixed in https://github.com/raspberrypi/linux/pull/2333/commits/4fd87fbe66417a4c44cf879617e3d1e9abf857f0.

6by9 reviewed (changes requested):
One memory leak, and a couple of minor nit picks, otherwise it looks OK on a first pass.
@popcornmix You're more aware of where the vc-sm cache handling ioctl's are used. Any comments?

@pelwell 4.9 or 4.14? I guess it needs to be picked across to the other one anyway.

pelwell commented:
As you say, it needs cherry-picking either way, so you might as well keep this PR against 4.9.

Terminus-IMRC commented:
Thank you for reviewing!!

Terminus-IMRC committed:
vcsm: Fix obscure conditions

Signed-off-by: Sugizaki Yukimasa <i.can.speak.c.and.basic@gmail.com>

Terminus-IMRC committed:
vcsm: Fix memory leaking on clean_invalid2 ioctl handler

Signed-off-by: Sugizaki Yukimasa <i.can.speak.c.and.basic@gmail.com>

Terminus-IMRC committed:
vcsm: Describe the use of cache operation constants

Signed-off-by: Sugizaki Yukimasa <i.can.speak.c.and.basic@gmail.com>

Terminus-IMRC committed:
vcsm: Fix obscure conditions again

Signed-off-by: Sugizaki Yukimasa <i.can.speak.c.and.basic@gmail.com>

popcornmix commented:
I temporarily added this PR to 4.14 last night to get some testing in the Milhouse LibreELEC nightly builds. Unfortunately a number of people have reported a regression of video playback randomly hanging on startup. These commits were isolated as introducing the problem:
https://forum.kodi.tv/showthread.php?tid=298461&pid=2689109#pid2689109

Terminus-IMRC commented:
Thanks. It may be because the func is flushing a page which is not yet mapped by fault handler, according to [this reply](https://forum.kodi.tv/showthread.php?tid=298461&pid=2689209#pid2689209). I will implement page table walk which may solve the problem.

popcornmix commented:
Yes, CLEAN_INVALID2 was an optimised API for framebuffers that we know are contiguous and pinned on allocation.
Other buffers may only have pages pinned when first accessed.
Flushing a non-pinned page does fault, and the page table walk code does avoid that (but with a certain performance cost).

Terminus-IMRC commented:
So my changes are to flush/clean/inv correctly on CLEAN_INVALID and other minor stuffs, rather than unifying them... (Anyway, I keep going)

Terminus-IMRC committed:
vcsm: Add no-op cache operation constant

Signed-off-by: Sugizaki Yukimasa <i.can.speak.c.and.basic@gmail.com>

Terminus-IMRC committed:
vcsm: Revert to do page-table-walk-based cache manipulating on some ioctl calls

On FLUSH, INVALID, CLEAN_INVALID ioctl calls, cache operations based on
page table walk were used in case that the buffer of the cache is not
pinned.  So reverted to do page-table-based cache manipulating.

Signed-off-by: Sugizaki Yukimasa <i.can.speak.c.and.basic@gmail.com>

Terminus-IMRC commented:
Reverted to operate cache by walking page table for FLUSH, INVALID, CLEAN_INVALID ioctls. This should work with LibreELEC......

Terminus-IMRC commented:
I confirmed it's now fixed with LibreELEC!

popcornmix commented:
Last night's LibreELEC build included update patch and seems fine.

Terminus-IMRC commented:
Thanks for merging! Closing.
